### PR TITLE
Feature/956 : API v2:  List all the entry points

### PR DESCRIPTION
**Pull request by jmcanterafonseca**
This is a Work in progress. Please do not review yet.

Thanks!

### CODE REVIEW

**Comment by jmcanterafonseca** [comment]
this is ready to be reviewed. @kzangeli r?

**Comment by kzangeli** [comment]
We'll do the #958 first.
After that I'll merge this branch with develop (with #958 merged into develop).

However, it might be good to comment on this PR untouched first.
Let's do that! :-)

File: src/app/contextBroker/contextBroker.cpp [hunk 1]
```diff
@@ -176,6 +176,8 @@
 #include "common/string.h"
 
 
+#include "serviceRoutines/entryPointsTreat.h"
```
Comment (kzangeli) [inline src/app/contextBroker/contextBroker.cpp]: Group with the other includes of files from 'lib/serviceRoutines'

File: src/app/contextBroker/contextBroker.cpp [hunk 2]
```diff
@@ -819,6 +826,10 @@ PaArgument paArgs[] =
   { "*",      CEAAT,   CEAAT_COMPS_V1,       "",               badVerbGetPostOnly                                  }
 
 
+#define SIMPLE_API_V2                                                                                                \
```
Comment (kzangeli) [inline src/app/contextBroker/contextBroker.cpp]: SIMPLE_API_V2 will be changed to CONVENIENCE_OPERATIONS_V2, following the older names (when merged with develop).

NTC

File: src/lib/serviceRoutines/entryPointsTreat.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+)
+{
+  std::string out     = "{";
+  std::string indent  = "";
+
+  out += valueTag(indent,        "entities_url",      ENTITIES_URL,      ciP->outFormat, true);
+  out += valueTag(indent + "  ", "types_url",         TYPES_URL,         ciP->outFormat, true);
+  out += valueTag(indent + "  ", "subscriptions_url", SUBSCRIPTIONS_URL, ciP->outFormat, true);
+  out += valueTag(indent + "  ", "registrations_url", REGISTRATIONS_URL, ciP->outFormat, false);
+
```
Comment (kzangeli) [inline src/lib/serviceRoutines/entryPointsTreat.cpp]: The functions in tag.cpp/h are meant for v1 (where JSON and XML are supported).
For v2 there are no helper functions yet and I'm not sure there will be any.
There functions add indents and newlines (that we don't want for v2), so please reimplement these four lines without  valueTag:

    out += "\"entities_url\":\"" + ENTITIES_URL + "\",";  ...

File: src/lib/serviceRoutines/entryPointsTreat.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+#include "rest/ConnectionInfo.h"
+#include "ngsi/ParseData.h"
+
+
+#define TYPES_URL            "/v2/types"
+#define ENTITIES_URL         "/v2/entities"
+#define SUBSCRIPTIONS_URL    "/v2/subscriptions"
+#define REGISTRATIONS_URL    "/v2/registrations"
+
```
Comment (kzangeli) [inline src/lib/serviceRoutines/entryPointsTreat.h]: Are these defines used outside entryPointsTreat.cpp?
If not, put them in entryPointsTreat.cpp instead.

File: src/app/contextBroker/contextBroker.cpp [hunk 5]
```diff
@@ -176,6 +176,8 @@
 #include "common/string.h"
```
Comment (fgalan) [inline src/app/contextBroker/contextBroker.cpp]: Entry in CHANGES_NEXT_RELEASE missing (I understand that referring to issue number https://github.com/telefonicaid/fiware-orion/issues/956)

File: src/lib/serviceRoutines/entryPointsTreat.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/serviceRoutines/entryPointsTreat.cpp]: Yeah, render v2 style is much more cleaner (eg. https://github.com/telefonicaid/fiware-orion/pull/958/files#diff-bd0a73ceaed2f25f0e58475c73512974R182)

File: test/functionalTest/cases/v2/retrieve_entry_point.test [hunk 6]
```diff
@@ -0,0 +1,54 @@
+# Copyright (c) 2015 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/v2/retrieve_entry_point.test]: The subdirectories within cases/ follow the pattern: NNN_XXXX where NNN is the number of the issue which functionality/fix is being tested and XXX is a descriptive string. In this case I suggest:

956_list_all_entry_points

File: test/functionalTest/cases/v2/retrieve_entry_point.test [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+echo
+
+--REGEXPECT--
+1: ++++++++++++++++++++
+HTTP/1.1 200 OK
+Content-Length: 153
+Content-Type: application/json
+Date: REGEX(.*)
+
+{
```
Comment (fgalan) [inline test/functionalTest/cases/v2/retrieve_entry_point.test]: One "philosofical" question... :)

Whey do we need this operation? I mean, the client programmer is supposed to know the URLs of the operations after reading the API documentation.

File: src/lib/serviceRoutines/entryPointsTreat.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/serviceRoutines/entryPointsTreat.cpp]: Comes without saying, but, merge in develop and use the new macros

File: test/functionalTest/cases/v2/retrieve_entry_point.test (hunk 7 above)
Comment (kzangeli) [inline test/functionalTest/cases/v2/retrieve_entry_point.test]: I had this idea (I still like it) to include a REST operation that returns the list of all operations accepted by the broker, including a short description, the in/out payloads and accepted URI params for each operation.

This is a little bit like that operation (we could build on it obtain what I had in mind).

It's like documentation builtin in the broker.

Comment (fgalan) [inline test/functionalTest/cases/v2/retrieve_entry_point.test]: @kzangeli I think that your idea is aready covered by an existing issue, isn't it?

**Comment by jmcanterafonseca** [comment]
@fgalan r?

File: src/app/contextBroker/contextBroker.cpp (hunk 5 above)
Comment (fgalan) [inline src/app/contextBroker/contextBroker.cpp]: Fixed.

File: test/functionalTest/cases/v2/retrieve_entry_point.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/v2/retrieve_entry_point.test]: Fixed

File: test/functionalTest/cases/v2/retrieve_entry_point.test (hunk 7 above)
Comment (fgalan) [inline test/functionalTest/cases/v2/retrieve_entry_point.test]: NTC

**Comment by fgalan** [comment]
LGTM

File: src/app/contextBroker/contextBroker.cpp (hunk 1 above)
Comment (jmcanterafonseca) [inline src/app/contextBroker/contextBroker.cpp]: FIXED

File: src/lib/serviceRoutines/entryPointsTreat.cpp (hunk 3 above)
Comment (jmcanterafonseca) [inline src/lib/serviceRoutines/entryPointsTreat.cpp]: FIXED

File: src/lib/serviceRoutines/entryPointsTreat.h (hunk 4 above)
Comment (jmcanterafonseca) [inline src/lib/serviceRoutines/entryPointsTreat.h]: I think they can potentially be used in other parts of the code so I would prefer to leave them in the header file

NTC

File: src/app/contextBroker/contextBroker.cpp [hunk 8]
```diff
@@ -330,6 +331,10 @@ PaArgument paArgs[] =
 //
 // /v2 API
 //
+
+#define EPS                EntryPointsRequest
+#define EPS_V2             1, { "v2"             }
```
Comment (kzangeli) [inline src/app/contextBroker/contextBroker.cpp]: These are the *components* of the request, so, pls change the name to **EPS_COMPS_V2** (you'll see that *all* other defines have COMPS in their name)

Comment (jmcanterafonseca) [inline src/app/contextBroker/contextBroker.cpp]: FIXED

**Comment by fgalan** [comment]
LGTM

**Pull request merged**